File: packages/backend-core/src/knowledge-read/domain/ranking.ts
/**
 * Knowledge-read bounded context — routing / ranking / scoring rules.
 *
 * Pure retrieval ranking rules (mode routing, candidate merge weights,
 * rerank boosts, graph merge, lexical score computation) with zero
 * framework, DB or I/O imports. The retrieval infrastructure and recall
 * coordinator render these rules over recall candidates.
 */

import type {
  BoundaryContext,
  BoundaryExplanation,
  FreshnessDecayConfig,
  GraphQueryRuntimeState,
  RetrievalQuery,
} from '@trapmap/contracts';
import { cosineSimilarity as sharedCosineSimilarity } from '@trapmap/lib';

import { type BoundaryEntryView, computeBoundaryScoreDelta } from './boundary.js';
import { type TokenMatchDetailLike, normalizeQuery } from './tokenization.js';

export const cosineSimilarity = sharedCosineSimilarity;

// ---------------------------------------------------------------------------
// Channel vocabulary
// ---------------------------------------------------------------------------

export const RECALL_CHANNEL_SEMANTIC = 'semantic' as const;
export const RECALL_CHANNEL_KEYWORD = 'keyword' as const;
export const RECALL_CHANNEL_GRAPH = 'graph' as const;

// ---------------------------------------------------------------------------
// Candidate shapes
// ---------------------------------------------------------------------------

export interface RecallCandidateLike<E> {
  entry: E;
  channel: string;
  score: number;
  tokenMatches: TokenMatchDetailLike[];
}

export interface MergedCandidateLike<E> {
  entry: E;
  semanticScore: number;
  keywordScore: number;
  graphScore?: number;
  channelScores: Record<string, number>;
  combinedScore: number;
  tokenMatches: TokenMatchDetailLike[];
  channels: string[];
  preRerankScore: number;
  finalScore: number;
  boundaryScoreDelta?: number;
  decayMultiplier?: number;
  boundaryExplanation?: BoundaryExplanation;
}

export function createSemanticCandidate<E>(entry: E, score: number): RecallCandidateLike<E> {
  return { entry, channel: RECALL_CHANNEL_SEMANTIC, score, tokenMatches: [] };
}

// ---------------------------------------------------------------------------
// Version-match decay
// ---------------------------------------------------------------------------

export interface VersionMatchMultiplierInput {
  /** Semver version declared by the artifact (absent for unversioned artifacts) */
  artifactVersion: string | null | undefined;
  /** Runtime versions from the query boundary context ({package, version} pairs) */
  queryVersions: ReadonlyArray<{ package: string; version: string }> | null | undefined;
  /** Decay freshness type from the entry decay metadata */
  freshnessType: string | null | undefined;
  /** Freshness decay configuration (consumed read-only) */
  decayConfig: FreshnessDecayConfig;
}

/**
 * Step multiplier for versioned decay: an artifact whose declared version
 * equals any query version keeps full weight; a versioned artifact that
 * declares a version not present in the query versions is down-weighted by
 * `mismatchMultiplier`. An artifact WITHOUT a declared version is treated as
 * neutral (1): "unknown" is not "mismatch", so it must not be silently
 * penalized while runtime version wiring is pending. Non-versioned freshness
 * types, queries without version constraints, and disabled or absent
 * versioned config all yield 1 (no decay).
 *
 * Matching is exact string equality between the artifact version and the
 * version field of the query's {package, version} entries (package names do
 * not participate in the comparison).
 */
export function versionMatchMultiplier(input: VersionMatchMultiplierInput): number {
  const { artifactVersion, queryVersions, freshnessType, decayConfig } = input;
  if (freshnessType !== 'versioned') return 1;
  if (!queryVersions || queryVersions.length === 0) return 1;
  const versioned = decayConfig.versioned;
  if (!versioned || versioned.enabled === false) return 1;
  if (artifactVersion === undefined || artifactVersion === null) return 1;
  const matched = queryVersions.some((entry) => entry.version === artifactVersion);
  return matched ? versioned.matchMultiplier : versioned.mismatchMultiplier;
}

// ---------------------------------------------------------------------------
// Mode routing
// ---------------------------------------------------------------------------

export interface RoutingDecisionLike {
  selectedMode: string;
  routeFamily: string;
  routingReason: string;
  fallbackApplied: boolean;
  fallbackTarget: string | null;
  confidenceScore: number | null;
  confidenceBucket: 'low' | 'medium' | 'high' | null;
  channelsPlanned: string[];
  channelsUsed: string[];
}

export function routingDecision(mode: string): RoutingDecisionLike {
  const channels =
    mode === 'graph-assisted'
      ? [RECALL_CHANNEL_SEMANTIC, RECALL_CHANNEL_KEYWORD, RECALL_CHANNEL_GRAPH]
      : mode === 'hybrid'
        ? [RECALL_CHANNEL_SEMANTIC, RECALL_CHANNEL_KEYWORD]
        : [RECALL_CHANNEL_SEMANTIC];
  return {
    selectedMode:
      mode === 'semantic'
        ? 'local'
        : mode === 'hybrid'
          ? 'hybrid'
          : mode === 'graph-assisted'
            ? 'mix'
            : 'local',
    routeFamily: 'entry',
    routingReason: 'explicit-mode',
    fallbackApplied: !['semantic', 'hybrid', 'graph-assisted'].includes(mode),
    fallbackTarget: null,
    confidenceScore: null,
    confidenceBucket: null,
    channelsPlanned: channels,
    channelsUsed: [],
  };
}

/** Validation message for an unknown query mode (mode whitelist rendering). */
export function buildUnknownModeMessage(mode: string, knownModes: readonly string[]): string {
  return `Invalid query mode: ${mode}. Must be one of: ${knownModes.join(', ')}`;
}

// ---------------------------------------------------------------------------
// Candidate merge / rerank
// ---------------------------------------------------------------------------

/** Weights applied when merging semantic and keyword candidates. */
export const MERGE_SEMANTIC_WEIGHT = 0.6;
export const MERGE_KEYWORD_WEIGHT = 0.4;

/** Rerank boosts and penalties. */
export const DUAL_CHANNEL_RERANK_BOOST = 0.15;
export const TOKEN_COVERAGE_BONUS = 0.1;
export const TOKEN_COVERAGE_RATIO = 0.5;
export const STALE_DECAY_PENALTY = 0.1;

export const DEFAULT_FRESHNESS_DECAY_CONFIG: FreshnessDecayConfig = {
  evergreen: { enabled: false },
  versioned: {
    enabled: true,
    mode: 'step',
    matchMultiplier: 1,
    mismatchMultiplier: 0.5,
  },
  volatile: {
    enabled: true,
    mode: 'exponential',
    halfLifeDays: 30,
    zeroDays: 90,
    floor: 0.3,
  },
};

export function mergeCandidates<E extends { id: string }>(
  semantic: RecallCandidateLike<E>[],
  keyword: RecallCandidateLike<E>[],
): MergedCandidateLike<E>[] {
  const candidates = new Map<string, MergedCandidateLike<E>>();
  for (const candidate of semantic)
    candidates.set(candidate.entry.id, {
      entry: candidate.entry,
      semanticScore: candidate.score,
      keywordScore: 0,
      graphScore: 0,
      channelScores: { [RECALL_CHANNEL_SEMANTIC]: candidate.score },
      combinedScore: candidate.score * MERGE_SEMANTIC_WEIGHT,
      tokenMatches: [],
      channels: [RECALL_CHANNEL_SEMANTIC],
      preRerankScore: candidate.score * MERGE_SEMANTIC_WEIGHT,
      finalScore: candidate.score * MERGE_SEMANTIC_WEIGHT,
    });
  for (const candidate of keyword) {
    const existing = candidates.get(candidate.entry.id);
    if (existing) {
      existing.keywordScore = candidate.score;
      existing.channelScores[RECALL_CHANNEL_KEYWORD] = candidate.score;
      existing.tokenMatches = candidate.tokenMatches;
      existing.channels = [RECALL_CHANNEL_SEMANTIC, RECALL_CHANNEL_KEYWORD];
      existing.combinedScore =
        existing.semanticScore * MERGE_SEMANTIC_WEIGHT + candidate.score * MERGE_KEYWORD_WEIGHT;
      existing.preRerankScore = existing.combinedScore;
      existing.finalScore = existing.combinedScore;
    } else
      candidates.set(candidate.entry.id, {
        entry: candidate.entry,
        semanticScore: 0,
        keywordScore: candidate.score,
        graphScore: 0,
        channelScores: { [RECALL_CHANNEL_KEYWORD]: candidate.score },
        combinedScore: candidate.score * MERGE_KEYWORD_WEIGHT,
        tokenMatches: candidate.tokenMatches,
        channels: [RECALL_CHANNEL_KEYWORD],
        preRerankScore: candidate.score * MERGE_KEYWORD_WEIGHT,
        finalScore: candidate.score * MERGE_KEYWORD_WEIGHT,
      });
  }
  return [...candidates.values()].sort(
    (left, right) =>
      right.combinedScore - left.combinedScore || left.entry.id.localeCompare(right.entry.id),
  );
}

export interface RerankableEntryView {
  decayMeta?: { decayState?: string } | null;
  boundary?: BoundaryEntryView | null;
}

export function rerankCandidates<E extends { id: string } & RerankableEntryView>(
  candidates: MergedCandidateLike<E>[],
  tokens: string[],
  options: {
    maxCandidates: number;
    boundaryContext?: BoundaryContext;
    freshnessConfig: FreshnessDecayConfig;
    earlyTerminationThreshold?: number;
  },
): MergedCandidateLike<E>[] {
  const topScore = Math.max(...candidates.map((candidate) => candidate.combinedScore));
  const threshold = options.earlyTerminationThreshold;
  const retained =
    threshold === undefined
      ? candidates
      : candidates.filter((candidate) => candidate.combinedScore >= topScore * threshold);
  return retained
    .map((candidate) => {
      const preRerankScore = candidate.combinedScore;
      let finalScore = preRerankScore;
      if (
        candidate.channels.includes(RECALL_CHANNEL_SEMANTIC) &&
        candidate.channels.includes(RECALL_CHANNEL_KEYWORD)
      )
        finalScore += DUAL_CHANNEL_RERANK_BOOST;
      if (
        tokens.length > 0 &&
        new Set(candidate.tokenMatches.map((match) => match.token)).size / tokens.length >=
          TOKEN_COVERAGE_RATIO
      )
        finalScore += TOKEN_COVERAGE_BONUS;
      if (candidate.entry.decayMeta?.decayState === 'stale') finalScore -= STALE_DECAY_PENALTY;
      finalScore += computeBoundaryScoreDelta(candidate.entry, options.boundaryContext);
      finalScore = Math.min(1, Math.max(0, finalScore));
      return {
        ...candidate,
        combinedScore: finalScore,
        preRerankScore,
        finalScore,
      };
    })
    .sort(
      (left, right) =>
        right.combinedScore - left.combinedScore || left.entry.id.localeCompare(right.entry.id),
    )
    .slice(0, options.maxCandidates);
}

// ---------------------------------------------------------------------------
// Graph merge
// ---------------------------------------------------------------------------

export const GRAPH_SCORE_BOOST_FACTOR = 0.2;

export function mergeCandidatesWithGraph<E extends { id: string }>(
  hybridMerged: MergedCandidateLike<E>[],
  graphCandidates: RecallCandidateLike<E>[],
): MergedCandidateLike<E>[] {
  const result = [...hybridMerged];

  for (const graphCandidate of graphCandidates) {
    const existing = result.find((c) => c.entry.id === graphCandidate.entry.id);

    if (existing) {
      existing.channels.push(RECALL_CHANNEL_GRAPH);
      existing.graphScore = graphCandidate.score;
      const preRerankScore = existing.combinedScore;
      const finalScore = Math.min(
        1,
        preRerankScore + graphCandidate.score * GRAPH_SCORE_BOOST_FACTOR,
      );
      existing.combinedScore = finalScore;
      existing.preRerankScore = preRerankScore;
      existing.finalScore = finalScore;
    } else {
      const score = graphCandidate.score;
      result.push({
        entry: graphCandidate.entry,
        semanticScore: 0,
        keywordScore: 0,
        graphScore: graphCandidate.score,
        channelScores: { [RECALL_CHANNEL_GRAPH]: graphCandidate.score },
        combinedScore: score,
        tokenMatches: [],
        channels: [RECALL_CHANNEL_GRAPH],
        preRerankScore: score,
        finalScore: score,
      });
    }
  }

  result.sort((a, b) => b.combinedScore - a.combinedScore);
  return result;
}

/** Routing channels actually used by a recall execution (semantic when empty). */
export function inferChannelsFromMerged(
  mergedCandidates?: Array<{ channels: string[] }> | null,
): string[] {
  if (!mergedCandidates || mergedCandidates.length === 0) {
    return [RECALL_CHANNEL_SEMANTIC];
  }
  const channelSet = new Set<string>();
  for (const candidate of mergedCandidates) {
    for (const ch of candidate.channels) {
      channelSet.add(ch);
    }
  }
  return Array.from(channelSet);
}

export interface GraphRecallTraceLike {
  mergeMode: 'mixed';
  graphExpansion: 'local-neighborhood';
  backendKind: GraphQueryRuntimeState['backendKind'];
  backendMode: GraphQueryRuntimeState['mode'];
  graphCandidateCount: number;
}

export function createGraphRecallTrace(
  runtimeState: GraphQueryRuntimeState | undefined,
  graphCandidateCount: number,
): GraphRecallTraceLike {
  return {
    mergeMode: 'mixed',
    graphExpansion: 'local-neighborhood',
    backendKind: runtimeState?.backendKind ?? 'memory',
    backendMode: runtimeState?.mode ?? 'disabled',
    graphCandidateCount,
  };
}

// ---------------------------------------------------------------------------
// Semantic scoring
// ---------------------------------------------------------------------------

export interface ScorableEntryView {
  labels: readonly string[];
  scope: string;
  shortcut: string;
  detail: string;
}

export function buildEmbeddingText(entry: ScorableEntryView): string {
  const labelsText = entry.labels.join(' ');
  return `${entry.shortcut}\n${entry.detail}\n${labelsText}`.trim();
}

export const LEXICAL_BOOST_FULL_MATCH = 0.55;
export const LEXICAL_BOOST_PER_TOKEN = 0.3;
export const LABEL_SCORE_BOOST = 0.05;
export const SCOPE_SCORE_BOOST = 0.03;

export function computeLexicalIntentBoost(seed: string, entry: ScorableEntryView): number {
  const queryTokens = normalizeQuery(seed);
  if (queryTokens.length === 0) return 0;

  const entryTokens = normalizeQuery(buildEmbeddingText(entry));
  if (entryTokens.length === 0) return 0;

  const overlapCount = queryTokens.filter((token) => entryTokens.includes(token)).length;
  if (overlapCount === 0) return 0;

  const ratio = overlapCount / queryTokens.length;
  const baseBoost = ratio >= 1 ? LEXICAL_BOOST_FULL_MATCH : ratio * LEXICAL_BOOST_PER_TOKEN;
  return Math.min(LEXICAL_BOOST_FULL_MATCH, baseBoost);
}

export function computeScore(
  similarity: number,
  entry: ScorableEntryView,
  filters: RetrievalQuery['filters'],
  seed?: string,
): number {
  let score = Math.max(0, Math.min(1, similarity));

  if (filters.labels.length > 0) {
    const matchingLabels = filters.labels.filter((label) => entry.labels.includes(label));
    const labelBoost = matchingLabels.length * LABEL_SCORE_BOOST;
    score = Math.min(1, score + labelBoost);
  }

  if (filters.scopes.length === 1 && filters.scopes[0] === entry.scope) {
    score = Math.min(1, score + SCOPE_SCORE_BOOST);
  }

  if (seed) {
    const lexicalBoost = computeLexicalIntentBoost(seed, entry);
    score = Math.min(1, score + lexicalBoost);
  }

  return score;
}

File: services/go-accelerator/internal/service/ranking/ranking.go
package ranking

import (
	"math"
	"sort"
	"strings"
)

const (
	MergeSemanticWeight      = 0.6
	MergeKeywordWeight       = 0.4
	DualChannelRerankBoost   = 0.08
	TokenCoverageRatio       = 0.5
	TokenCoverageBonus       = 0.05
	StaleDecayPenalty        = 0.1
	GraphScoreBoostFactor    = 0.2
)

// Entry mirrors MergedCandidateLike minimal fields for batch ranking.
type Entry struct {
	ID                string             `json:"id"`
	SemanticScore     float64            `json:"semanticScore"`
	KeywordScore      float64            `json:"keywordScore"`
	GraphScore        *float64           `json:"graphScore,omitempty"`
	ChannelScores     map[string]float64 `json:"channelScores"`
	CombinedScore     float64            `json:"combinedScore"`
	TokenMatches      []TokenMatch       `json:"tokenMatches"`
	Channels          []string           `json:"channels"`
	PreRerankScore    float64            `json:"preRerankScore"`
	FinalScore        float64            `json:"finalScore"`
	BoundaryScoreDelta *float64          `json:"boundaryScoreDelta,omitempty"`
	DecayMultiplier   *float64           `json:"decayMultiplier,omitempty"`
	Labels            []string           `json:"labels"`
	Scope             string             `json:"scope"`
	Shortcut          string             `json:"shortcut"`
	Detail            string             `json:"detail"`
	DecayState        *string            `json:"decayState,omitempty"`
	Boundary          *Boundary          `json:"boundary,omitempty"`
}

type TokenMatch struct {
	Token  string   `json:"token"`
	Fields []string `json:"fields"`
}

type Boundary struct {
	Context   []string    `json:"context,omitempty"`
	Exclusions []Exclusion `json:"exclusions,omitempty"`
}

type Exclusion struct {
	Kind        string `json:"kind"`
	Description string `json:"description"`
}

type BoundaryContext struct {
	Contexts []string `json:"contexts"`
	Platform *string  `json:"platform,omitempty"`
}

func normalizeBoundaryLabel(label string) string {
	lower := strings.ToLower(label)
	// replace whitespace with '-'
	var b strings.Builder
	prevDash := false
	for _, r := range lower {
		if r == ' ' || r == '\t' || r == '\n' {
			if !prevDash {
				b.WriteRune('-')
				prevDash = true
			}
		} else if (r >= 'a' && r <= 'z') || (r >= '0' && r <= '9') || r == '-' {
			b.WriteRune(r)
			prevDash = (r == '-')
		} else {
			// drop other chars like original replace /[^a-z0-9-]/g
		}
	}
	s := b.String()
	if len(s) > 64 {
		s = s[:64]
	}
	return s
}

func computeBoundaryScoreDelta(entry Entry, ctx *BoundaryContext) float64 {
	if ctx == nil || entry.Boundary == nil {
		return 0
	}
	delta := 0.0
	for _, qc := range ctx.Contexts {
		delta += contextScoreDelta(entry, qc)
	}
	if ctx.Platform != nil {
		delta += platformScoreDelta(entry, *ctx.Platform)
	}
	return delta
}

func contextScoreDelta(entry Entry, queryContext string) float64 {
	normalized := normalizeBoundaryLabel(queryContext)
	delta := 0.0
	if entry.Boundary != nil {
		for _, ex := range entry.Boundary.Exclusions {
			if ex.Kind == "context" {
				lowerDesc := strings.ToLower(ex.Description)
				if strings.Contains(lowerDesc, normalized) || strings.Contains(lowerDesc, strings.ToLower(queryContext)) {
					delta -= 0.15
					break
				}
			}
		}
		for _, label := range entry.Boundary.Context {
			if normalizeBoundaryLabel(label) == normalized {
				delta += 0.1
				break
			}
		}
	}
	return delta
}

func platformScoreDelta(entry Entry, platform string) float64 {
	if entry.Boundary == nil {
		return 0
	}
	lowerPlat := strings.ToLower(platform)
	for _, ex := range entry.Boundary.Exclusions {
		if ex.Kind == "platform" && strings.Contains(strings.ToLower(ex.Description), lowerPlat) {
			return -0.15
		}
	}
	return 0
}

// MergeCandidates mirrors backend-core mergeCandidates (weights 0.6/0.4)
func MergeCandidates(semantic, keyword []Entry) []Entry {
	m := make(map[string]*Entry)
	for _, c := range semantic {
		cc := c
		cc.ChannelScores = map[string]float64{"semantic": c.SemanticScore}
		cc.CombinedScore = c.SemanticScore * MergeSemanticWeight
		cc.PreRerankScore = cc.CombinedScore
		cc.FinalScore = cc.CombinedScore
		cc.Channels = []string{"semantic"}
		// normalize fields
		if cc.TokenMatches == nil {
			cc.TokenMatches = []TokenMatch{}
		}
		m[cc.ID] = &cc
	}
	for _, c := range keyword {
		if existing, ok := m[c.ID]; ok {
			existing.KeywordScore = c.SemanticScore // keyword score stored in SemanticScore slot for input
			// Actually keyword input's score is in SemanticScore when using Entry uniformly
			// Prefer explicit KeywordScore if provided
			ks := c.KeywordScore
			if ks == 0 {
				ks = c.SemanticScore
			}
			existing.KeywordScore = ks
			if existing.ChannelScores == nil {
				existing.ChannelScores = map[string]float64{}
			}
			existing.ChannelScores["keyword"] = ks
			existing.TokenMatches = c.TokenMatches
			existing.Channels = []string{"semantic", "keyword"}
			existing.CombinedScore = existing.SemanticScore*MergeSemanticWeight + ks*MergeKeywordWeight
			existing.PreRerankScore = existing.CombinedScore
			existing.FinalScore = existing.CombinedScore
		} else {
			cc := c
			ks := c.KeywordScore
			if ks == 0 {
				ks = c.SemanticScore
			}
			cc.KeywordScore = ks
			cc.SemanticScore = 0
			cc.ChannelScores = map[string]float64{"keyword": ks}
			cc.CombinedScore = ks * MergeKeywordWeight
			cc.PreRerankScore = cc.CombinedScore
			cc.FinalScore = cc.CombinedScore
			cc.Channels = []string{"keyword"}
			if cc.TokenMatches == nil {
				cc.TokenMatches = []TokenMatch{}
			}
			m[cc.ID] = &cc
		}
	}
	out := make([]Entry, 0, len(m))
	for _, v := range m {
		out = append(out, *v)
	}
	sort.Slice(out, func(a, b int) bool {
		if out[a].CombinedScore != out[b].CombinedScore {
			return out[a].CombinedScore > out[b].CombinedScore
		}
		return out[a].ID < out[b].ID
	})
	return out
}

func contains(slice []string, s string) bool {
	for _, v := range slice {
		if v == s {
			return true
		}
	}
	return false
}

// RerankCandidates mirrors backend-core rerankCandidates
func RerankCandidates(candidates []Entry, tokens []string, maxCandidates int, ctx *BoundaryContext) []Entry {
	if len(candidates) == 0 {
		return candidates
	}
	topScore := candidates[0].CombinedScore
	for _, c := range candidates {
		if c.CombinedScore > topScore {
			topScore = c.CombinedScore
		}
	}
	// Early termination threshold not exposed in batch API yet; keep all
	retained := candidates
	// Token coverage bonus prep
	tokenSetSize := len(tokens)
	out := make([]Entry, 0, len(retained))
	for _, c := range retained {
		pre := c.CombinedScore
		final := pre
		if contains(c.Channels, "semantic") && contains(c.Channels, "keyword") {
			final += DualChannelRerankBoost
		}
		if tokenSetSize > 0 {
			uniq := make(map[string]bool)
			for _, tm := range c.TokenMatches {
				uniq[tm.Token] = true
			}
			if float64(len(uniq))/float64(tokenSetSize) >= TokenCoverageRatio {
				final += TokenCoverageBonus
			}
		}
		if c.DecayState != nil && *c.DecayState == "stale" {
			final -= StaleDecayPenalty
		}
		final += computeBoundaryScoreDelta(c, ctx)
		final = math.Max(0, math.Min(1, final))
		cc := c
		cc.PreRerankScore = pre
		cc.CombinedScore = final
		cc.FinalScore = final
		// also set BoundaryScoreDelta for observability
		bd := computeBoundaryScoreDelta(c, ctx)
		cc.BoundaryScoreDelta = &bd
		out = append(out, cc)
	}
	sort.Slice(out, func(a, b int) bool {
		if out[a].CombinedScore != out[b].CombinedScore {
			return out[a].CombinedScore > out[b].CombinedScore
		}
		return out[a].ID < out[b].ID
	})
	if maxCandidates > 0 && len(out) > maxCandidates {
		out = out[:maxCandidates]
	}
	return out
}

func MergeCandidatesWithGraph(hybrid []Entry, graph []Entry) []Entry {
	result := make([]Entry, len(hybrid))
	copy(result, hybrid)
	// map for fast find
	idxByID := make(map[string]int, len(result))
	for i, c := range result {
		idxByID[c.ID] = i
	}
	for _, g := range graph {
		if idx, ok := idxByID[g.ID]; ok {
			existing := &result[idx]
			existing.Channels = append(existing.Channels, "graph")
			if existing.ChannelScores == nil {
				existing.ChannelScores = map[string]float64{}
			}
			gs := g.SemanticScore
			if g.GraphScore != nil {
				gs = *g.GraphScore
			}
			if g.KeywordScore != 0 {
				gs = g.KeywordScore
			}
			existing.GraphScore = &gs
			existing.ChannelScores["graph"] = gs
			pre := existing.CombinedScore
			final := math.Min(1, pre+gs*GraphScoreBoostFactor)
			existing.PreRerankScore = pre
			existing.CombinedScore = final
			existing.FinalScore = final
		} else {
			gs := g.SemanticScore
			if g.GraphScore != nil {
				gs = *g.GraphScore
			}
			// ensure GraphScore set
			ng := gs
			newEntry := g
			newEntry.SemanticScore = 0
			newEntry.KeywordScore = 0
			newEntry.GraphScore = &ng
			newEntry.ChannelScores = map[string]float64{"graph": gs}
			newEntry.CombinedScore = gs
			newEntry.PreRerankScore = gs
			newEntry.FinalScore = gs
			newEntry.Channels = []string{"graph"}
			if newEntry.TokenMatches == nil {
				newEntry.TokenMatches = []TokenMatch{}
			}
			result = append(result, newEntry)
			idxByID[newEntry.ID] = len(result) - 1
		}
	}
	sort.Slice(result, func(a, b int) bool { return result[a].CombinedScore > result[b].CombinedScore })
	return result
}

// ComputeScore mirrors backend-core computeScore
func ComputeScore(similarity float64, labels []string, scope string, filtersLabels []string, filtersScopes []string, seed string, entryLabels []string, entryScope string, entryShortcut string, entryDetail string) float64 {
	score := math.Max(0, math.Min(1, similarity))
	if len(filtersLabels) > 0 {
		matching := 0
		for _, fl := range filtersLabels {
			for _, el := range labels {
				if fl == el {
					matching++
					break
				}
			}
		}
		score = math.Min(1, score+float64(matching)*0.05)
	}
	if len(filtersScopes) == 1 && filtersScopes[0] == scope {
		score = math.Min(1, score+0.03)
	}
	if seed != "" {
		// lexical boost
		queryTokens := tokenize(seed)
		entryText := entryShortcut + "\n" + entryDetail + "\n" + strings.Join(entryLabels, " ")
		entryTokens := tokenize(entryText)
		if len(queryTokens) > 0 && len(entryTokens) > 0 {
			overlap := 0
			entrySet := make(map[string]bool, len(entryTokens))
			for _, t := range entryTokens {
				entrySet[t] = true
			}
			for _, t := range queryTokens {
				if entrySet[t] {
					overlap++
				}
			}
			if overlap > 0 {
				ratio := float64(overlap) / float64(len(queryTokens))
				base := ratio * 0.3
				if ratio >= 1 {
					base = 0.55
				}
				if base > 0.55 {
					base = 0.55
				}
				score = math.Min(1, score+base)
			}
		}
	}
	return score
}

func tokenize(text string) []string {
	// mirrors tokenization normalizeQuery: split on non-alnum, lower, dedup length>=2 not needed for lexical boost
	lower := strings.ToLower(text)
	parts := strings.FieldsFunc(lower, func(r rune) bool {
		return !(r >= 'a' && r <= 'z') && !(r >= '0' && r <= '9')
	})
	set := make(map[string]bool)
	out := []string{}
	for _, p := range parts {
		if p == "" {
			continue
		}
		if !set[p] {
			set[p] = true
			out = append(out, p)
		}
	}
	return out
}
